src/Test.UnitTests.BinSkim.Driver/CompilerDataLoggerTests.cs

```diff
@@ -166,32 +175,179 @@ public void CompilerDataLogger_Dispose_ShouldReadSarifV1()
+            fileSystem.Verify(fileSystem => fileSystem.FileOpenRead(sarifLogPath), Times.Never);
+            sendItems.Count.Should().Be(0);
+            context.Dispose();
+        }
 
+        [Fact]
+        public void CompilerDataLogger_ShouldInitializeFromEnvironmentVariable()
+        {
+            CompilerDataLogger.s_injectedTelemetryClient = null;
+            CompilerDataLogger.s_injectedTelemetryConfiguration = null;
+            BinaryAnalyzerContext context = CreateTestContext();
+            var fileSystem = new Mock<IFileSystem>();
+            string randomString = Guid.NewGuid().ToString();
+            Environment.SetEnvironmentVariable("BinskimCompilerDataAppInsightsKey", randomString);
+            var sendItems = new List<ITelemetry>();
             context.Policy = new Sarif.PropertiesDictionary();
-            logger = new CompilerDataLogger(sarifLogFilePath, sarifVersion, context ?? new BinaryAnalyzerContext(), fileSystem);
+            var logger = new CompilerDataLogger(SarifPath, Sarif.SarifVersion.Current, context, fileSystem.Object);
+
+            logger.Should().NotBeNull();
+            Environment.SetEnvironmentVariable("BinskimCompilerDataAppInsightsKey", null);
+            context.Dispose();
+        }
+
+        [Fact]
+        public void CompilerDataLogger_WriteException_ShouldLogExceptionMessageString()
+        {
+            var fileSystem = new Mock<IFileSystem>();
+            BinaryAnalyzerContext context = CreateTestContext();
+            List<ITelemetry> sendItems = TestSetup(SarifPath, context, Sarif.SarifVersion.Current, out CompilerDataLogger compilerDataLogger, fileSystem.Object);
+            compilerDataLogger.WriteException(context, "testException");
+            sendItems.Count.Should().Be(1);
+            context.Dispose();
+        }
+
+        [Fact]
+        public void CompilerDataLogger_WriteException_ShouldNotLogExceptionMessage_WhenTelemetryClientIsNull()
+        {
+            var fileSystem = new Mock<IFileSystem>();
+            BinaryAnalyzerContext context = CreateTestContext();
+            List<ITelemetry> sendItems = TestSetup(SarifPath, context, Sarif.SarifVersion.Current, out CompilerDataLogger compilerDataLogger, fileSystem.Object, true);
+            compilerDataLogger.WriteException(context, "testException");
+            sendItems.Count.Should().Be(0);
+            context.Dispose();
+        }
+
+        [Fact]
+        public void CompilerDataLogger_WriteException_ShouldLogExecutionExceptionFromBaseLineTestSarif()
+        {
+            string sarifLogPath = Path.Combine(PEBinaryTests.BaselineTestDataDirectory, ExpectedFolder, SampleSarifPath);
+            var fileSystem = new Mock<IFileSystem>();
+            BinaryAnalyzerContext context = CreateTestContext();
+            List<ITelemetry> sendItems = TestSetup(sarifLogPath, context, Sarif.SarifVersion.Current, out CompilerDataLogger compilerDataLogger, fileSystem.Object);
```

Comment:
Investigate further collapsing common instantiation and setup into helper method(s). ex: construct filesystem mock in testsetup if not null